### PR TITLE
perf(gatsby): stop using sets and use ordered arrays instead

### PR DESCRIPTION
**Pull request by pvdz**
Turns out that Sets do not scale nicely. In one benchmark that was using a query with two filters ($eq and $gte), it turned out that creating a `Set` for each `$gte` was taking up 75% of the time and scaled dramatically.

So I flopped things around. No more creating of Sets. Instead there's a "manual" intersection that is O(n) and also dedupes. Still room for improvement, but this is a first step.

Note: the first commit will contain the juicy bits (all tests pass on this). The second part is only renaming and changing comments. Tried to eliminate all mentions of "set" and move them to array where that made sense.

### CODE REVIEW

**Comment by jonniebigodes** [comment]
Enviado do Correio<https://go.microsoft.com/fwlink/?LinkId=550986> para Windows 10

________________________________
De: Peter van der Zee <notifications@github.com>
Enviado: Tuesday, May 26, 2020 4:40:19 PM
Para: gatsbyjs/gatsby <gatsby@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Assunto: [gatsbyjs/gatsby] perf(gatsby): stop using sets and use ordered arrays instead (#24486)


Turns out that Sets do not scale nicely. In one benchmark that was using a query with two filters ($eq and $gte), it turned out that creating a Set for each $gte was taking up 75% of the time and scaled dramatically.

So I flopped things around. No more creating of Sets. Instead there's a "manual" intersection that is O(n) and also dedupes. Still room for improvement, but this is a first step.

________________________________
You can view, comment on, or merge this pull request online at:

  https://github.com/gatsbyjs/gatsby/pull/24486

Commit Summary

  *   perf(gatsby): stop using sets and use ordered arrays instead

File Changes

  *   M packages/gatsby/src/redux/nodes.ts<https://github.com/gatsbyjs/gatsby/pull/24486/files#diff-df09f60e4f190103964a9c69bb44801d> (156)
  *   M packages/gatsby/src/redux/run-sift.js<https://github.com/gatsbyjs/gatsby/pull/24486/files#diff-5d2176cb542d5c01d884139970be33fb> (134)
  *   M packages/gatsby/src/schema/__tests__/__snapshots__/kitchen-sink.js.snap<https://github.com/gatsbyjs/gatsby/pull/24486/files#diff-af306b46ac913d90a7e1416c4cfb8b95> (12)
  *   M packages/gatsby/src/schema/__tests__/connection-filter-on-linked-nodes.js<https://github.com/gatsbyjs/gatsby/pull/24486/files#diff-fdde9ff9c89ddced5724f81b70f2882c> (4)

Patch Links:

  *   https://github.com/gatsbyjs/gatsby/pull/24486.patch
  *   https://github.com/gatsbyjs/gatsby/pull/24486.diff

**Comment by pvdz** [comment]
Posted a followup wip based on this branch where slices are prevented by working directlly on ranges: https://github.com/gatsbyjs/gatsby/pull/24527

This, together with dropping sift in https://github.com/gatsbyjs/gatsby/pull/24345 , should complete the sift work for now.

File: packages/gatsby/src/redux/nodes.ts [hunk 1]
```diff
@@ -688,29 +698,32 @@ export const getNodesFromCacheByValue = (
...
         return a < b ? -1 : a > b ? 1 : 0
       }) // Just sort to preserve legacy order as much as possible.
       .forEach((v: FilterValueNullable) =>
-        filterCache.byValue.get(v)?.forEach(v => set.add(v))
+        filterCache.byValue.get(v)?.forEach(v => arr.push(v))
       )
 
-    return set
+    // TODO: proper merge sort
+    return arr.sort((A, B) => (A.id < B.id ? -1 : A.id === B.id ? 0 : 1))
```
Comment (vladar) [inline packages/gatsby/src/redux/nodes.ts:726]: Do we still need to sort `filterValueArr` above if we re-sort the resulting `arr` here?

File: packages/gatsby/src/redux/nodes.ts [hunk 2]
```diff
@@ -761,7 +776,9 @@ export const getNodesFromCacheByValue = (
       if (cache) cache.forEach(node => set.delete(node))
     }
 
-    return set
+    // TODO: there's probably a more efficient algorithm to do set
+    //       subtraction in such a way that we dont have to resort here
+    return [...set].sort((A, B) => (A.id < B.id ? -1 : A.id === B.id ? 0 : 1))
```
Comment (vladar) [inline packages/gatsby/src/redux/nodes.ts:781]: Looks like `nin` and `ne` have a similar logic for diffing lists of nodes. Maybe we can extract it to a separate utility like `diffNodes` similar to `intersectNodes`?

File: packages/gatsby/src/redux/nodes.ts (hunk 1 above)
Comment (pvdz) [reply packages/gatsby/src/redux/nodes.ts:726]: No, good point :)

File: packages/gatsby/src/redux/nodes.ts (hunk 2 above)
Comment (pvdz) [reply packages/gatsby/src/redux/nodes.ts:781]: Good point :)

**Comment by pvdz** [comment]
Ok, I changed it to use the `counter` instead, which should preserve insertion sort.

The tests were updated (because our framework doesn't enforce the `.counter` property and so some tests don't bother to add it). I've manually checked the changed snapshots that insertion sort holds.

**Pull request merged**